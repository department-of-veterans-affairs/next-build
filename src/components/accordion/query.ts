import { QueryFormatter } from 'next-drupal-query'
import { ParagraphAccordion } from '@/types/drupal/paragraph'
import { AccordionItem } from '@/components/accordion/formatted-type'
import { getHtmlFromField } from '@/lib/utils/getHtmlFromField'

export const formatter: QueryFormatter<ParagraphAccordion, AccordionItem> = (
  entity: ParagraphAccordion
) => {
  return {
    type: entity.type as AccordionItem['type'],
    id: entity.id,
    entityId: entity.drupal_internal__id,
    header: entity.field_header || null,
    html: getHtmlFromField(entity.field_rich_wysiwyg),
  }
}
