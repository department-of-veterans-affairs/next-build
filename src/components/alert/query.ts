import { QueryFormatter } from 'next-drupal-query'
import { ParagraphAlert } from '@/types/drupal/paragraph'
import { Alert, AlertBlock, AlertType } from '@/components/alert/formatted-type'
import { formatParagraph } from '@/lib/drupal/paragraphs'
import { queries } from '@/lib/drupal/queries'

export const formatter: QueryFormatter<ParagraphAlert, Alert> = (
  entity: ParagraphAlert
) => {
  return {
    type: entity.type as Alert['type'],
    id: entity.id ?? null,
    entityId: entity.drupal_internal__id ?? null,
    alertType: entity.field_alert_type as AlertType,
    heading: entity.field_alert_heading,
    blockReference: queries.formatData(
      'block--alert',
      entity.field_alert_block_reference
    ) as AlertBlock,
    paragraphs: entity.field_va_paragraphs.map((paragraph) =>
      formatParagraph(paragraph)
    ),
  }
}
