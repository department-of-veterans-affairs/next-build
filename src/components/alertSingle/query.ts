import { QueryFormatter, QueryParams } from 'next-drupal-query'
import { DrupalJsonApiParams } from 'drupal-jsonapi-params'
import { ParagraphAlertSingle } from '@/types/drupal/paragraph'
import {
  AlertSingle,
  AlertBlock,
  AlertNonReusable,
} from '@/types/formatted/alert'
import { formatParagraph } from '@/lib/drupal/paragraphs'
import { queries } from '@/data/queries'

export const params: QueryParams<null> = () => {
  return new DrupalJsonApiParams().addInclude([
    'field_alert_block_reference',
    'field_alert_block_reference.field_alert_content',
    'field_alert_non_reusable_ref',
    'field_alert_non_reusable_ref.field_va_paragraphs',
  ])
}

export const formatter: QueryFormatter<ParagraphAlertSingle, AlertSingle> = (
  entity: ParagraphAlertSingle
) => {
  return {
    type: entity.type as AlertSingle['type'],
    id: entity.id,
    entityId: entity.drupal_internal__id,
    alertSelection:
      entity.field_alert_selection as AlertSingle['alertSelection'],
    blockReference: queries.formatData(
      'block--alert',
      entity.field_alert_block_reference
    ) as AlertBlock,
    nonReusableRef: formatParagraph(
      entity.field_alert_non_reusable_ref
    ) as AlertNonReusable,
  }
}
