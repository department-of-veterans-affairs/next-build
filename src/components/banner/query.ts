import { QueryData, QueryFormatter } from 'next-drupal-query'
import { BANNER_RESOURCE_TYPES } from '@/lib/constants/resourceTypes'
import { BannersData } from './formatted-type'
import { drupalClient } from '@/lib/drupal/drupalClient'
import { getHtmlFromField } from '@/lib/utils/getHtmlFromField'

export type BannerDataOpts = {
  itemPath?: string
}

// The banner data endpoint is a custom endpoint provided by Drupal due to how banners are associated with a page.
// A given page does not reference a banner node via entity reference, banner node types have a field that lists what
// paths they are supposed to be visible on. This endpoint queries banners based on their path lists.
// See docroot/modules/custom/va_gov_api/src/Controller/BannerAlertsController.php in va.gov-cms for more info.
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const data: QueryData<BannerDataOpts, any> = async (opts) => {
  if (opts.itemPath) {
    const bannerUrl = `${process.env.NEXT_PUBLIC_DRUPAL_BASE_URL}/api/v1/banner-alerts?path=${opts.itemPath}`

    const response = await drupalClient.fetch(bannerUrl)
    const json = await response.json()
    return json.data
  }

  return []
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const formatter: QueryFormatter<any, BannersData> = (entities) => {
  return entities?.map((banner) => {
    switch (banner?.type.target_id as string) {
      case BANNER_RESOURCE_TYPES.BASIC:
        // this field returns 'perm' or 'dismiss' string instead of bool
        const dismiss = banner.field_dismissible_option === 'dismiss'

        return {
          id: banner.nid,
          title: banner.title,
          body: getHtmlFromField(banner.body),
          alertType: banner.field_alert_type,
          dismiss,
          type: banner.type.target_id,
        }
      case BANNER_RESOURCE_TYPES.PROMO:
        return {
          id: banner.nid,
          title: banner.title,
          href: banner.field_link?.uri,
          alertType: banner.field_promo_type,
          type: banner.type.target_id,
        }
      default:
        return null
    }
  })
}
