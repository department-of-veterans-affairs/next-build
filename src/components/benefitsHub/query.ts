import { QueryData, QueryFormatter, QueryParams } from 'next-drupal-query'
import { DrupalJsonApiParams } from 'drupal-jsonapi-params'
import { drupalClient } from '@/lib/drupal/drupalClient'
import { NodeLandingPage } from '@/types/drupal/node'
import { BenefitsHub } from './formatted-type'
import { RESOURCE_TYPES } from '@/lib/constants/resourceTypes'
import { ExpandedStaticPropsContext } from '@/lib/drupal/staticProps'
import {
  entityBaseFields,
  fetchSingleEntityOrPreview,
} from '@/lib/drupal/query'
import { getNestedIncludes } from '@/lib/utils/queries'
import { formatter as formatListOfLinkTeasers } from '@/components/listOfLinkTeasers/query'

// Define the query params for fetching node--landing_page for benefits hub.
export const params: QueryParams<null> = () => {
  return new DrupalJsonApiParams().addInclude([
    'field_spokes',
    'field_spokes.field_va_paragraphs',
  ])
}

// Define the option types for the data loader.
export type BenefitsHubDataOpts = {
  id: string
  context?: ExpandedStaticPropsContext
}

// Implement the data loader.
export const data: QueryData<BenefitsHubDataOpts, NodeLandingPage> = async (
  opts
): Promise<NodeLandingPage> => {
  const entity = (await fetchSingleEntityOrPreview(
    opts,
    RESOURCE_TYPES.BENEFITS_HUB,
    params
  )) as NodeLandingPage

  return entity
}

export const formatter: QueryFormatter<NodeLandingPage, BenefitsHub> = (
  entity: NodeLandingPage
) => {
  // Only return field_spokes if it's the correct type (ParagraphListOfLinkTeasers)
  const fieldSpokes = Array.isArray(entity.field_spokes)
    ? entity.field_spokes.filter(
        (spoke) => spoke.type === 'paragraph--list_of_link_teasers'
      )
    : []

  return {
    ...entityBaseFields(entity),
    title: entity.title,
    fieldIntroText: entity.field_intro_text,
    hubLabel: entity.field_home_page_hub_label,
    teaserText: entity.field_teaser_text,
    titleIcon: entity.field_title_icon,
    fieldSpokes: fieldSpokes,
  }
}
