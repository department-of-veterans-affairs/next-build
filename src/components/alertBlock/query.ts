// Define the query params for fetching block--alert.
import { BlockAlert } from '@/types/drupal/block'
import { QueryFormatter, QueryParams } from 'next-drupal-query'
import { AlertBlock, AlertType } from '@/components/alert/formatted-type'
import { formatParagraph } from '@/lib/drupal/paragraphs'
import { Wysiwyg } from '@/components/wysiwyg/formatted-type'
import { ExpandableText } from '@/components/expandableText/formatted-type'
import { DrupalJsonApiParams } from 'drupal-jsonapi-params'

export const params: QueryParams<null> = () =>
  new DrupalJsonApiParams().addInclude(['field_alert_content'])

export const formatter: QueryFormatter<BlockAlert, AlertBlock> = (
  entity: BlockAlert
) => {
  if (!entity) {
    return null
  }

  return {
    alertType: (entity.field_alert_type === 'information'
      ? 'info'
      : entity.field_alert_type) as AlertType,
    id: entity.id,
    title: entity.field_alert_title,
    content: formatParagraph(entity.field_alert_content) as
      | Wysiwyg
      | ExpandableText,
    // {
    //   header:
    //     entity.field_alert_content.field_text_expander ??
    //     entity.field_alert_heading,
    //   text: entity.field_alert_content?.field_wysiwyg?.processed ?? '',
    // },
  }
}
