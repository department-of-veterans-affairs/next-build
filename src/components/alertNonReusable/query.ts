// Define the query params for fetching block--alert.
import { QueryFormatter } from 'next-drupal-query'
import { AlertType, AlertNonReusable } from '@/components/alert/formatted-type'
import { ParagraphNonReusableAlert } from '@/types/drupal/paragraph'
import { formatParagraph } from '@/lib/drupal/paragraphs'

export const formatter: QueryFormatter<
  ParagraphNonReusableAlert,
  AlertNonReusable
> = (entity: ParagraphNonReusableAlert) => {
  if (!entity) {
    return null
  }

  return {
    type: entity.type as AlertNonReusable['type'],
    id: entity.id,
    entityId: entity.drupal_internal__id,
    alertType: (entity.field_alert_type === 'information'
      ? 'info'
      : entity.field_alert_type) as AlertType,
    heading: entity.field_alert_heading,
    paragraphs: entity.field_va_paragraphs.map((paragraph) => formatParagraph(paragraph)),
  }
}
