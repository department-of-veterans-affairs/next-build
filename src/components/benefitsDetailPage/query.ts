import { QueryData, QueryFormatter, QueryParams } from 'next-drupal-query'
import { DrupalJsonApiParams } from 'drupal-jsonapi-params'
import { NodeBenefitsDetailPage } from '@/types/drupal/node'
import { BenefitsDetailPage } from './formatted-type'
import {
  RESOURCE_TYPES,
  // PARAGRAPH_RESOURCE_TYPES,
} from '@/lib/constants/resourceTypes'
import { ExpandedStaticPropsContext } from '@/lib/drupal/staticProps'
import {
  entityBaseFields,
  fetchSingleEntityOrPreview,
} from '@/lib/drupal/query'
import { formatter as formatListOfLinkTeasers } from '@/components/listOfLinkTeasers/query'
import { formatter as formatAdministration } from '@/components/administration/query'
import { formatter as formatAlertBlock } from '@/components/alertBlock/query'
import { formatParagraph } from '@/lib/drupal/paragraphs'
// import { getNestedIncludes } from '@/lib/utils/queries'
import { getHtmlFromField } from '@/lib/utils/getHtmlFromField'

// Define the query params for fetching node--page (benefits detail page).
export const params: QueryParams<null> = () => {
  return new DrupalJsonApiParams().addInclude([
    'field_administration',
    'field_alert',
    'field_related_links',
    'field_related_links.field_va_paragraphs',
    // ...getNestedIncludes('field_content_block', PARAGRAPH_RESOURCE_TYPES.QA),
    // ...getNestedIncludes('field_featured_content', PARAGRAPH_RESOURCE_TYPES.QA),
  ])
}

// Define the option types for the data loader.
export type BenefitsDetailPageDataOpts = {
  id: string
  context?: ExpandedStaticPropsContext
}

// Implement the data loader.
export const data: QueryData<
  BenefitsDetailPageDataOpts,
  NodeBenefitsDetailPage
> = async (opts): Promise<NodeBenefitsDetailPage> => {
  const entity = (await fetchSingleEntityOrPreview(
    opts,
    RESOURCE_TYPES.BENEFITS_DETAIL_PAGE,
    params
  )) as NodeBenefitsDetailPage

  return entity
}

export const formatter: QueryFormatter<
  NodeBenefitsDetailPage,
  BenefitsDetailPage
> = (entity: NodeBenefitsDetailPage) => {
  // TODO: There is an ownership problem with the related links for /education/benefit-rates/post-9-11-gi-bill-rates/past-rates-2021-2022.
  // This needs to be resolved in Drupal, but for now I'm just going to not show the related links
  // if it fails to fetch it.
  let relatedLinks = null
  if (
    entity.field_related_links &&
    entity.field_related_links.field_va_paragraphs
  ) {
    relatedLinks = formatListOfLinkTeasers(entity.field_related_links)
  }

  return {
    ...entityBaseFields(entity),
    title: entity.title,
    description: entity.field_description || null,
    introText: getHtmlFromField(entity.field_intro_text_limited_html) || null,
    showTableOfContents: entity.field_table_of_contents_boolean ?? false,
    alert: entity.field_alert ? formatAlertBlock(entity.field_alert) : null,
    // featuredContent:
    //   entity.field_featured_content?.map((paragraph) =>
    //     formatParagraph(paragraph)
    //   ) || null,
    // contentBlock:
    //   entity.field_content_block?.map((paragraph) =>
    //     formatParagraph(paragraph)
    //   ) || null,
    relatedLinks,
    administration: entity.field_administration
      ? formatAdministration(entity.field_administration)
      : null,
  }
}
