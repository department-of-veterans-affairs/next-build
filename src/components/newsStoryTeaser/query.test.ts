/**
 * @jest-environment node
 */

import { NodeNewsStory } from '@/types/drupal/node'
import mockData from '@/components/newsStory/mock.json'
import { RESOURCE_TYPES } from '@/lib/constants/resourceTypes'
import { params, formatter } from './query'

const nodeNewsStoryMock: NodeNewsStory = mockData

describe(`${RESOURCE_TYPES.STORY}Teaser formatData`, () => {
  test('outputs formatted data', () => {
    const formattedData = formatter(nodeNewsStoryMock)
    expect(formattedData).toMatchSnapshot()
  })

  test('handles missing or null fields correctly', () => {
    const modifiedMock = {
      ...nodeNewsStoryMock,
      field_media: null,
      field_listing: undefined,
    }

    const formattedData = formatter(modifiedMock)
    expect(formattedData.image).toBeNull()
    expect(formattedData.link).toBe(modifiedMock.path.alias)
  })
})

describe('DrupalJsonApiParams configuration for newsStoryTeaser', () => {
  test('params function sets the correct include fields', () => {
    const paramsInstance = params()
    const queryString = decodeURIComponent(paramsInstance.getQueryString())
    expect(queryString).toMatch(
      /include=field_media,field_media.image,field_listing/
    )
  })
})
