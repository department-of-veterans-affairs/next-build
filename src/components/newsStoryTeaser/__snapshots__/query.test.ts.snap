// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`node--news_storyTeaser formatData outputs formatted data 1`] = `
{
  "headingLevel": "h2",
  "id": "bb4dcfc1-736b-4a9e-aa48-23c2dd4a4980",
  "image": {
    "alt": null,
    "height": 304,
    "id": "c85b7882-f035-4b55-8b71-fcf81fade350",
    "links": {
      "1_1_square_medium_thumbnail": {
        "href": "http://content-build-medc0xjkxm4jmpzxl3tfbcs7qcddsivh.ci.cms.va.gov/sites/default/files/styles/1_1_square_medium_thumbnail/public/hub_promos/health-care.png",
        "meta": {
          "linkParams": {
            "height": "240",
            "rel": [
              "drupal://jsonapi/extensions/consumer_image_styles/links/relation-types/#derivative",
            ],
            "width": "240",
          },
          "rel": [
            "drupal://jsonapi/extensions/consumer_image_styles/links/relation-types/#derivative",
          ],
        },
      },
      "2_1_large": {
        "href": "http://content-build-medc0xjkxm4jmpzxl3tfbcs7qcddsivh.ci.cms.va.gov/sites/default/files/styles/2_1_large/public/hub_promos/health-care.png",
        "meta": {
          "linkParams": {
            "height": "512",
            "rel": [
              "drupal://jsonapi/extensions/consumer_image_styles/links/relation-types/#derivative",
            ],
            "width": "1024",
          },
          "rel": [
            "drupal://jsonapi/extensions/consumer_image_styles/links/relation-types/#derivative",
          ],
        },
      },
      "2_1_medium_thumbnail": {
        "href": "http://content-build-medc0xjkxm4jmpzxl3tfbcs7qcddsivh.ci.cms.va.gov/sites/default/files/styles/2_1_medium_thumbnail/public/hub_promos/health-care.png",
        "meta": {
          "linkParams": {
            "height": "240",
            "rel": [
              "drupal://jsonapi/extensions/consumer_image_styles/links/relation-types/#derivative",
            ],
            "width": "480",
          },
          "rel": [
            "drupal://jsonapi/extensions/consumer_image_styles/links/relation-types/#derivative",
          ],
        },
      },
      "2_3_medium_thumbnail": {
        "href": "http://content-build-medc0xjkxm4jmpzxl3tfbcs7qcddsivh.ci.cms.va.gov/sites/default/files/styles/2_3_medium_thumbnail/public/hub_promos/health-care.png",
        "meta": {
          "linkParams": {
            "height": "480",
            "rel": [
              "drupal://jsonapi/extensions/consumer_image_styles/links/relation-types/#derivative",
            ],
            "width": "320",
          },
          "rel": [
            "drupal://jsonapi/extensions/consumer_image_styles/links/relation-types/#derivative",
          ],
        },
      },
      "3_2_medium_thumbnail": {
        "href": "http://content-build-medc0xjkxm4jmpzxl3tfbcs7qcddsivh.ci.cms.va.gov/sites/default/files/styles/3_2_medium_thumbnail/public/hub_promos/health-care.png",
        "meta": {
          "linkParams": {
            "height": "320",
            "rel": [
              "drupal://jsonapi/extensions/consumer_image_styles/links/relation-types/#derivative",
            ],
            "width": "480",
          },
          "rel": [
            "drupal://jsonapi/extensions/consumer_image_styles/links/relation-types/#derivative",
          ],
        },
      },
      "7_2_medium_thumbnail": {
        "href": "http://content-build-medc0xjkxm4jmpzxl3tfbcs7qcddsivh.ci.cms.va.gov/sites/default/files/styles/7_2_medium_thumbnail/public/hub_promos/health-care.png",
        "meta": {
          "linkParams": {
            "height": "300",
            "rel": [
              "drupal://jsonapi/extensions/consumer_image_styles/links/relation-types/#derivative",
            ],
            "width": "1050",
          },
          "rel": [
            "drupal://jsonapi/extensions/consumer_image_styles/links/relation-types/#derivative",
          ],
        },
      },
      "full_content_width": {
        "href": "http://content-build-medc0xjkxm4jmpzxl3tfbcs7qcddsivh.ci.cms.va.gov/sites/default/files/styles/full_content_width/public/hub_promos/health-care.png",
        "meta": {
          "linkParams": {
            "height": "304",
            "rel": [
              "drupal://jsonapi/extensions/consumer_image_styles/links/relation-types/#derivative",
            ],
            "width": "456",
          },
          "rel": [
            "drupal://jsonapi/extensions/consumer_image_styles/links/relation-types/#derivative",
          ],
        },
      },
      "large": {
        "href": "http://content-build-medc0xjkxm4jmpzxl3tfbcs7qcddsivh.ci.cms.va.gov/sites/default/files/styles/large/public/hub_promos/health-care.png",
        "meta": {
          "linkParams": {
            "height": "304",
            "rel": [
              "drupal://jsonapi/extensions/consumer_image_styles/links/relation-types/#derivative",
            ],
            "width": "456",
          },
          "rel": [
            "drupal://jsonapi/extensions/consumer_image_styles/links/relation-types/#derivative",
          ],
        },
      },
      "self": {
        "href": "https://content-build-medc0xjkxm4jmpzxl3tfbcs7qcddsivh.ci.cms.va.gov/jsonapi/file/file/e638fd17-9090-442b-a724-5113133b4d0f",
      },
      "viewport_width": {
        "href": "http://content-build-medc0xjkxm4jmpzxl3tfbcs7qcddsivh.ci.cms.va.gov/sites/default/files/styles/viewport_width/public/hub_promos/health-care.png",
        "meta": {
          "linkParams": {
            "height": "304",
            "rel": [
              "drupal://jsonapi/extensions/consumer_image_styles/links/relation-types/#derivative",
            ],
            "width": "456",
          },
          "rel": [
            "drupal://jsonapi/extensions/consumer_image_styles/links/relation-types/#derivative",
          ],
        },
      },
    },
    "title": null,
    "type": "media--image",
    "width": 456,
  },
  "introText": "When a hospital has a host of great doctors, honoring just two every year is challenging. ",
  "lastUpdated": "2019-05-14T15:35:12+00:00",
  "link": "/pittsburgh-health-care/stories/we-honor-outstanding-doctors",
  "published": true,
  "title": "We honor outstanding doctors",
  "type": "node--news_story",
}
`;
